### PR TITLE
Make 0 as invalid value for `min_children` in `has_child` query

### PR DESCRIPTION
**Pull request by codebird**
Just updated the test in file 

modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java

from `< 0` to `<= 0` and changed the error message.

### CODE REVIEW

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java
Comment (cbuescher) [inline modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: Would you mind changing that to simply say "positive"?

**Review by cbuescher — CHANGES_REQUESTED** [review]
Hi @codebird, thanks for the PR. I left a minor comment around the error message. Also at least the unit test HasChildQueryBuilderTests needs adapting to the new illegal values and message. Maybe other tests are failing. You can take a look at the TESTING.txt file if you are unsure of how to run the tests. Since the change is in the "server" subproject, I think running "./gradlew -p server check" from the root folder should catch most test failures. Otherwise we'll catch them when running the final CI tests. Let me know if you have any questions.

**Comment by cbuescher** [comment]
This relates to #32949.

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java
Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: on it @cbuescher

**Comment by cbuescher** [comment]
Another quick hint: apart from changing the min/max value cases in HasChildQueryBuilderTests#testIllegalValues, the #doCreateTestQueryBuilder method in the same test class will need some changes so we don't draw 0 values for min/max when randomly setting up the query builder.

File: modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java
Comment (codebird) [reply modules/parent-join/src/main/java/org/elasticsearch/join/query/HasChildQueryBuilder.java]: updated @cbuescher

**Comment by cbuescher** [comment]
@codebird the change looks good from what I see, unfortunately the diff is a bit large now with the unrelated changes merged in from master. I'm not sure this is a Github quirk, but could you try squashing your branch down to one commit and force-push that to your PR branch? That would make it easier to read I think. I will kick of the larger CI test suite after that. Thanks.

**Comment by codebird** [comment]
done @cbuescher  Thanks for the help

**Comment by cbuescher** [comment]
Something is weird. Your branch still shows changes in unrelated classes (e.g. TestingConventionsTasks.java), but when I look at the current state of that class on master it looks like that version is already similar with the one on your PR branch. Maybe a Github glitch. I will need to look into this more deeply before merging, but will run CI for now.
@elasticmachine test this please

**Comment by codebird** [comment]
that's because I did

```
git fetch upstream
git rebase upstream/master
```
Shouldn't I do this next time?

**Comment by codebird** [comment]
I update DEFAULT_MIN_CHILDREN to be 1 not 0, for the case where min_children is not provided at all, that's what was causing CI to fail.

**Comment by cbuescher** [comment]
@codebird I hope you don't mind but I merged in master once more to fix a merge conflict. Now the diff also looks okay (only two files changed). Will run the full tests again.
@elasticmachine test this please

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by cbuescher** [comment]
@elasticmachine test this please

**Comment by codebird** [comment]
@cbuescher I think you to re-initiate the tests. if I mention elasticmachine will it run the tests?

**Comment by cbuescher** [comment]
Unfortunately not. You need to be member of the "Elastic" organization to run the CI bot so our testing infrastructure cannot be DoS'ed. The current ci/1 failure seems unrelated to this change, I will just run that part again.
@elasticmachine run elasticsearch-ci/1

**Comment by codebird** [comment]
cute this elasticmachine :) Yeah I know it is not related to us, that's why I said you need to re-initiate

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by cbuescher** [comment]
The last ci/1 run died a while ago with "java.lang.OutOfMemoryError: Java heap space" https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/12487/console
Kicking of another one.
@elasticmachine run elasticsearch-ci/1

**Comment by codebird** [comment]
Failed again with errors not to do with this

**Comment by cbuescher** [comment]
Unfortunately. Just for reference, the last ones were timeouts: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/12510/
@elasticmachine run elasticsearch-ci/1

**Comment by codebird** [comment]
hmmm weird, it failed again  

`Failed to start elasticsearch: timed out after 30 seconds`

Just for my info does this always happen like this? Or something causing it now, systems are busy or something?

**Comment by cbuescher** [comment]
@codebird normally things pass more smoothly. We had some recent changes in the build system that might make this one a bit more bumpy. The latest failures are unrelated as well and look transient (some googleapis 500 Internal Server Errors in GCE plugin tests). This is unfortunate, but we want to have at least one clear run of the entire suite before merging anything, even if its only a few lines changed like this one.
@elasticmachine run elasticsearch-ci/1

**Comment by cbuescher** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by cbuescher** [comment]
@codebird ^ latest one is gce failure again, feels like a general problem in the testing infra at the moment (not just this PR). I'm going to wait on internal feedback before continuing re-runs here.

**Comment by cbuescher** [comment]
Merged master, maybe that helps.
@elasticmachine test this please

**Comment by cbuescher** [comment]
Merged master, maybe that helps.
@elasticmachine test this please

**Comment by codebird** [comment]
Wooooo it worked

**Review by cbuescher — APPROVED** [review]
Today is the lucky day ;-)
LGTM and will merge to master now, probably backport to 7.1 after that.

**Comment by codebird** [comment]
Thanks for the help @cbuescher. And sorry for the headache :P

**Comment by cbuescher** [comment]
No problem, and thanks for picking this up, much appreciated. I would like to backport this to 7.1 but I see problems with this since it would mean introducing a slightly different behaviour for the `has_child` query on a minor release. We usually don't introduce breaking changes like this in minors unless it fixes some serious bugs. This is more of a safety enhancement, so I think it should only go to master.

**Comment by codebird** [comment]
Maybe we should start with a warning on 7.1 just to prepare for this to go live some day. I don't know what are your rules for these warnings...

**Comment by cbuescher** [comment]
That would be a nice follow up indeed. Are you interested in doing that in a small PR against the 7.x branch? Look at the way DeprecationLogger is used e.g. in TermsQueryBuilder. Adding something to the "minMaxChildren" method and log a short notice for 0 values (that this will be rejected in 8.0 but if using 1 this should behave the same). In that case we'd also need to adapt tests in a similar way like in this PR because triggering the warning will cause tests to fail. In additon we'd need a test that checks that the 0 value triggers the warning. See e.g. TermsQueryBuilderTests#testTypeField how this can be done using the "assertWarnings" method.
If you are not interested in this I can open an issue regardless to follow up on it.

**Comment by codebird** [comment]
Yes interested since this is not urgent and a bit busy at work. Let's do it

**Comment by cbuescher** [comment]
Great, I opened https://github.com/elastic/elasticsearch/issues/41548 just so we don't forget following up on it. Feel free to grab it and let me know if there are any questions.

**Pull request merged**
